### ISSUE TITLE
Java bindings not just for Android

### ISSUE THREAD

**Issue by GommeAntiLegit** [mozilla/DeepSpeech#2129]
I tried using DeepSpeech in my Java project, but linking fails at runtime because the natives are only made for Android, as it seems. Are there any plans for making Java bindings with natives for all relevant platforms such as Windows, Linux, and MacOS?
Or even better: Is there any way how I could build natives for Linux x64 myself? 
Is there anything I need to pay attention to, to also have CUDA support?

Thank you for your help.

**Comment by lissyx** [comment]
Well, current bindings are really targetting Android because we have no use outside of Android for Java so far. Supporting other should not be impossible, but it requires you write proper packaging as JAR I guess. We have no plans for that yet, but if you can provide PR it's welcome.

Also, it's possible there might be some stuff that is currently really only okay under Android, but nothing not fixable.

>  Or even better: Is there any way how I could build natives for Linux x64 myself?
I guess it's just a matter of building current bindings as a JAR and properly loading `libdeepspeech.so`, for which you can rely on prebuilt ones.

>  Is there anything I need to pay attention to, to also have CUDA support?
On Java side, no idea. If you pick the CUDA enabled `libdeepspeech.so` it should work. But we don't have that on macOS though.

**Comment by lissyx** [comment]
@GommeAntiLegit So here is the gradle file for Android: https://github.com/mozilla/DeepSpeech/blob/685a0dbb894e3b7e68435cbab3aae34cdfed90ce/native_client/java/build.gradle

I guess we would need something similar for non android platforms? Though I have to admit I'm not knowledgeable a lot regarding Java world, so that's also why we have not spent too much time on that.

**Comment by GommeAntiLegit** [comment]
As it turns out, libdeepspeech.jni.so is also a required native library, which only works on Android because it requires libraries such as liblog.so, which are part of the Android NDK.

**Comment by lissyx** [comment]
> As it turns out, libdeepspeech.jni.so is also a required native library, which only works on Android because it requires libraries such as liblog.so, which are part of the Android NDK.

Yes, but it's a by-product of the Gradle build system for Android. I don't know how all of that should be handled for non-Android Java platforms.

**Comment by GommeAntiLegit** [comment]
I'm creating new Java bindings using the JNI. I have successfully linked the Java native function to the implementation in the shared library, but when DeepSpeech functions are called natively, Java reports an unknown symbol at runtime.

`/usr/lib/jvm/java-8-oracle/bin/java: symbol lookup error: /mnt/E08026E98026C63E/LIN_SSD_EXT/Development/Java/deep-speech-test/libdeepspeech-jni.so: undefined symbol: DS_PrintVersions`

The function invoked by Java is: `Java_org_mozilla_deepspeech_NativeImpl_printVersions` which then calls `DS_PrintVersions`.

`nm -D ./libdeepspeech-jni.so` prints:

```
0000000000201028 B __bss_start
                 w __cxa_finalize
                 U DS_PrintVersions
0000000000201028 D _edata
0000000000201030 B _end
0000000000000624 T _fini
                 w __gmon_start__
00000000000004e8 T _init
                 w _ITM_deregisterTMCloneTable
                 w _ITM_registerTMCloneTable
000000000000060a T Java_org_mozilla_deepspeech_NativeImpl_printVersions
```
`DS_PrintVersions` is unimplemented in this library, as I declared it as an extern function. The function is implemented in `libdeepspeech.so` from `native_client.amd64.cuda.linux.tar.xz` version `v0.5.0-alpha.10`.

`nm -D ./libdeepspeech.so` states that the following functions (and also many more) are defined:

```
0000000000cd19e0 T DS_CreateModel
0000000000cd19b0 T DS_DestroyModel
0000000000ccd770 T DS_DiscardStream
0000000000ccd5e0 T DS_EnableDecoderWithLM
0000000000cd15e0 T DS_FeedAudioContent
0000000000cd16b0 T DS_FinishStream
0000000000cd1700 T DS_FinishStreamWithMetadata
0000000000ccd7d0 T DS_FreeMetadata
0000000000ccd850 T DS_FreeString
0000000000ccde20 T DS_IntermediateDecode
0000000000ccd860 T DS_PrintVersions
0000000000ccef50 T DS_SetupStream
0000000000cd16e0 T DS_SpeechToText
0000000000cd1730 T DS_SpeechToTextWithMetadata
```

I'm loading both native libraries from Java:

```
System.load(new File("libdeepspeech.so").getAbsolutePath());
System.load(new File("libdeepspeech-jni.so").getAbsolutePath());
NativeImpl.printVersions();
```

Execution ends with exit code 127 and this is print to the error stream: `/usr/lib/jvm/java-8-oracle/bin/java: symbol lookup error: /mnt/E08026E98026C63E/LIN_SSD_EXT/Development/Java/deep-speech-test/libdeepspeech-jni.so: undefined symbol: DS_PrintVersions`

The C++ function that is invoked here is:
```
JNIEXPORT void JNICALL Java_org_mozilla_deepspeech_NativeImpl_printVersions(JNIEnv *, jclass) {
    DS_PrintVersions();
}
```

I'm using CMake to build this shared library "libdeepspeech-jni.so".

This is my CMakeLists.txt:
```
cmake_minimum_required(VERSION 3.13)
project(deepspeech-jni)

set(CMAKE_CXX_STANDARD 14)

add_library(deepspeech-jni SHARED org_mozilla_deepspeech_NativeImpl.h impl/native_impl.cpp deepspeech/deepspeech.h)

set_target_properties(deepspeech-jni PROPERTIES LINKER_LANGUAGE CXX)

link_directories(${PROJECT_SOURCE_DIR}/lib)
link_libraries(deepspeech-jni deepspeech)
```
`org_mozilla_deepspeech_NativeImpl.h` contains the method prototypes that Java calls.

`NativeImp.printVersions()` is defined as

```
JNIEXPORT void JNICALL Java_org_mozilla_deepspeech_NativeImpl_printVersions(JNIEnv *, jclass);
```

The implementation of this prototype is present in `native_impl.cpp`.
`deepspeech/deepspeech.h` contains all prototypes for extern functions that are present in `libdeepspeech.so` such as `extern void DS_PrintVersions();`

I'm also linking the shared library against the `libdeepspeech.so`.

I can't really figure out why `DS_PrintVersions` is not found.

Thank you for your help.

**Comment by lissyx** [comment]
> ### EDIT: THIS ISSUE IS RESOLVED. IF YOU HAVE A SIMILAR ISSUE, SCROLL DOWN TO THE "EDIT:" HEADER
> 
> I'm creating new Java bindings using the JNI. I have successfully linked the Java native function to the implementation in the shared library, but when DeepSpeech functions are called natively, Java reports an unknown symbol at runtime.
> 
> `/usr/lib/jvm/java-8-oracle/bin/java: symbol lookup error: /mnt/E08026E98026C63E/LIN_SSD_EXT/Development/Java/deep-speech-test/libdeepspeech-jni.so: undefined symbol: DS_PrintVersions`
> 
> The function invoked by Java is: `Java_org_mozilla_deepspeech_NativeImpl_printVersions` which then calls `DS_PrintVersions`.
> 
> `nm -D ./libdeepspeech-jni.so` prints:
> 
> ```
> 0000000000201028 B __bss_start
>                  w __cxa_finalize
>                  U DS_PrintVersions
> 0000000000201028 D _edata
> 0000000000201030 B _end
> 0000000000000624 T _fini
>                  w __gmon_start__
> 00000000000004e8 T _init
>                  w _ITM_deregisterTMCloneTable
>                  w _ITM_registerTMCloneTable
> 000000000000060a T Java_org_mozilla_deepspeech_NativeImpl_printVersions
> ```
> 
> `DS_PrintVersions` is unimplemented in this library, as I declared it as an extern function. The function is implemented in `libdeepspeech.so` from `native_client.amd64.cuda.linux.tar.xz` version `v0.5.0-alpha.10`.
> 
> `nm -D ./libdeepspeech.so` states that the following functions (and also many more) are defined:
> 
> ```
> 0000000000cd19e0 T DS_CreateModel
> 0000000000cd19b0 T DS_DestroyModel
> 0000000000ccd770 T DS_DiscardStream
> 0000000000ccd5e0 T DS_EnableDecoderWithLM
> 0000000000cd15e0 T DS_FeedAudioContent
> 0000000000cd16b0 T DS_FinishStream
> 0000000000cd1700 T DS_FinishStreamWithMetadata
> 0000000000ccd7d0 T DS_FreeMetadata
> 0000000000ccd850 T DS_FreeString
> 0000000000ccde20 T DS_IntermediateDecode
> 0000000000ccd860 T DS_PrintVersions
> 0000000000ccef50 T DS_SetupStream
> 0000000000cd16e0 T DS_SpeechToText
> 0000000000cd1730 T DS_SpeechToTextWithMetadata
> ```
> 
> I'm loading both native libraries from Java:
> 
> ```
> System.load(new File("libdeepspeech.so").getAbsolutePath());
> System.load(new File("libdeepspeech-jni.so").getAbsolutePath());
> NativeImpl.printVersions();
> ```
> 
> Execution ends with exit code 127 and this is print to the error stream: `/usr/lib/jvm/java-8-oracle/bin/java: symbol lookup error: /mnt/E08026E98026C63E/LIN_SSD_EXT/Development/Java/deep-speech-test/libdeepspeech-jni.so: undefined symbol: DS_PrintVersions`
> 
> The C++ function that is invoked here is:
> 
> ```
> JNIEXPORT void JNICALL Java_org_mozilla_deepspeech_NativeImpl_printVersions(JNIEnv *, jclass) {
>     DS_PrintVersions();
> }
> ```
> 
> I'm using CMake to build this shared library "libdeepspeech-jni.so".
> 
> This is my CMakeLists.txt:
> 
> ```
> cmake_minimum_required(VERSION 3.13)
> project(deepspeech-jni)
> 
> set(CMAKE_CXX_STANDARD 14)
> 
> add_library(deepspeech-jni SHARED org_mozilla_deepspeech_NativeImpl.h impl/native_impl.cpp deepspeech/deepspeech.h)
> 
> set_target_properties(deepspeech-jni PROPERTIES LINKER_LANGUAGE CXX)
> 
> link_directories(${PROJECT_SOURCE_DIR}/lib)
> link_libraries(deepspeech-jni deepspeech)
> ```
> 
> `org_mozilla_deepspeech_NativeImpl.h` contains the method prototypes that Java calls.
> 
> `NativeImp.printVersions()` is defined as
> 
> ```
> JNIEXPORT void JNICALL Java_org_mozilla_deepspeech_NativeImpl_printVersions(JNIEnv *, jclass);
> ```
> 
> The implementation of this prototype is present in `native_impl.cpp`.
> `deepspeech/deepspeech.h` contains all prototypes for extern functions that are present in `libdeepspeech.so` such as `extern void DS_PrintVersions();`
> 
> I'm also linking the shared library against the `libdeepspeech.so`.
> 
> I can't really figure out why `DS_PrintVersions` is not found.
> 
> Thank you for your help.
> ### EDIT:
> 
> The library was linked incorrectly in my CMakeLists.txt.
> This is fixed now:
> 
> ```
> cmake_minimum_required(VERSION 3.13)
> project(deepspeech-jni)
> 
> set(CMAKE_CXX_STANDARD 14)
> 
> add_library(deepspeech-jni SHARED org_mozilla_deepspeech_NativeImpl.h impl/native_impl.cpp deepspeech/deepspeech.h)
> 
> add_library(deepspeech SHARED IMPORTED)
> set_target_properties( deepspeech PROPERTIES IMPORTED_LOCATION ${CMAKE_SOURCE_DIR}/lib/libdeepspeech.so )
> 
> target_link_libraries(deepspeech-jni deepspeech)
> ```

Should we assume it means it works properly, indeed? That's awesome news!

**Comment by GommeAntiLegit** [comment]
I'm currently implementing the JNI functions that call the deepspeech functions. printVersions is already tested and printed: 
"TensorFlow: v1.13.1-10-g3e0cc53
DeepSpeech: v0.5.0-alpha.10-0-g685a0db"
So, there is no reason why the other functions shouldn't work. I will make a pull request when I get it working.

**Comment by lissyx** [comment]
Yeah, that sounds like a lot of what is already done in `native_client/java/libdeepspeech/src/main/java/org/mozilla/deepspeech/libdeepspeech/`, and we have nothing Android-specific there, can you make sure you're not reinventing the wheel?

**Comment by reuben** [comment]
Could this be related to the C++ name mangling we recently fixed?

-- reuben

**Comment by reuben** [comment]
Nevermind, I didn't see that the DS_PrintVersions problems were fixed.

-- reuben

**Comment by GommeAntiLegit** [comment]
A rewrite of the java bindings is more than needed. Java naming conventions are neglected, they do not free native memory at all, so, many STT calls result in a gigantic memory leak. Sometimes native memory is allocated (not freed) that is not even used once before the pointer is overridden with an address to newly allocated memory. Loading of the native library must be more configurable or you end up with many cheap hacks in your code just to link the Cuda libraries. With more library loading options Android compatibility could be achieved with the same bindings. The bindings lack documentation or partially even parameter names (var1, var2, var3). The bindings are full of TODOS and comments stating "We should have something better than this". I'm currently rewriting them because I'm used to clean native libraries such as LWJGL which provide great documentation and insight what parameter types actually mean on the native side. eg usage of a long to represent a pointer or the usage of a IntBuffer to represent a call by reference integer. The bindings are using Java Arrays that are passed to JNI functions. This is not efficient at all as java array contents must be copied to be accessed natively as Java Objects do not have a static memory location and can be moved around by the Garage collector. Java NIO buffers should be used instead as they are a wrapper around native memory. The native memory is freed when the wrapper object is garbage collected. The memory of those buffers can be directly accessed by the functions of DeepSpeech without any conversion that results in a loss in performance. This is very important as audio data is often very large and copying of millions of bytes just to be freed again for nothing is not very efficient. On top of the pure bindings will also be an object oriented layer of abstraction similar to what is already present.

**Comment by lissyx** [comment]
> A rewrite of the java bindings is more than needed.

Maybe. My point is that we don't want to have to maintain Android **and** non-Android bindings ...

**Comment by lissyx** [comment]
@GommeAntiLegit Yes, well, you understand why we stated this was still **experimental** and why I told you we are not fluent in Java. We know a lot of things are not perfect, but there should not be such big leaks as you describe, @reuben carefully checked that we were properly freeing with `DS_FreeString` as well as `DS_FreeMetadata`, and using on-device did not show leaking.

But yeah, there is a lot to improve, so your feedback and fixes are more than welcome.

**Comment by lissyx** [comment]
@GommeAntiLegit Since we are close to making a 0.5.0 release, I'd like to know if you have any idea how much time it might require you to make things better? This feels like a lot of changes, with risks of regressions, I'd like to avoid pressuring you into doing it "in time".

**Comment by GommeAntiLegit** [comment]
> @GommeAntiLegit Since we are close to making a 0.5.0 release, I'd like to know if you have any idea how much time it might require you to make things better? This feels like a lot of changes, with risks of regressions, I'd like to avoid pressuring you into doing it "in time".

How much time is it until the release? I do have a few exams coming up in the next few days, but I might be able to do it in time, depending how much time there is left.


> > A rewrite of the java bindings is more than needed.
> 
> Maybe. My point is that we don't want to have to maintain Android **and** non-Android bindings ...

Yes it should be one library.

> @GommeAntiLegit Yes, well, you understand why we stated this was still **experimental** and why I told you we are not fluent in Java. We know a lot of things are not perfect, but there should not be such big leaks as you describe, @reuben carefully checked that we were properly freeing with `DS_FreeString` as well as `DS_FreeMetadata`, and using on-device did not show leaking.
> 
> But yeah, there is a lot to improve, so your feedback and fixes are more than welcome.

Yes, I overlooked this as my IDE displayed the method as unused, as the "finalize" method which calls the method is not used itself but invoked by the Garbage collector.

**Comment by lissyx** [comment]
> How much time is it until the release? I do have a few exams coming up in the next few days, but I might be able to do it in time, depending how much time there is left.

We don't have a strict date. If you have exams, they should be your top priority. My point was more to re-assure you that we value your feedback and contribution, and to make sure we can either wait for you if it's a quick change, or that you don't feel sending PR for nothing if we decide to go ahead and release soonish. Lots of people are waiting on 0.5.0 because we changed the model.

If the changes do no introduce any backward incompatibility, we could easily pick that in a 0.5.1 I guess.

**Comment by GommeAntiLegit** [comment]
As my exams are over, I started work on the Java bindings again. The C++ library is fully accessible in Java. Yet, there is no additional layer of abstraction to make this a little bit more pleasant to use in Java. To use the model of v0.4.1 I tried to re-export it from the checkpoint, but that just resulted in a crash. It seems as if there is some backwards incompatibility with older checkpoints. Someone claimed to have a fix for this and posted a modified version of a function of the python script, but using that function I got a model that I could load, but `speechToText` just returns an empty string no matter what. I was able to reproduce this with both my Java bindings and using the unchanged DeepSpeech C++ functions, so, I think that the model that was exported is corrupted. Something that supports that theory is that the v0.5.0-alpha.10 native client for Linux fails to load the model and exits with a RuntimeError. where the Trace-back points to `ds = Model(args.model, N_FEATURES, N_CONTEXT, args.alphabet, BEAM_WIDTH)`. Is there any other pre-trained model available that I can use for the newest version? Training is nearly impossible on my Laptop, so I rely on pre-trained models to properly test the bindings. It does not need to be any good at all. It should just produce some output so I can test the bindings.

**Comment by lissyx** [comment]
> As my exams are over, I started work on the Java bindings again. The C++ library is fully accessible in Java. Yet, there is no additional layer of abstraction to make this a little bit more pleasant to use in Java. To use the model of v0.4.1 I tried to re-export it from the checkpoint, but that just resulted in a crash. It seems as if there is some backwards incompatibility with older checkpoints. Someone claimed to have a fix for this and posted a modified version of a function of the python script, but using that function I got a model that I could load, but `speechToText` just returns an empty string no matter what. I was able to reproduce this with both my Java bindings and by calling the unchanged DeepSpeech functions from C++, so, I think that the model that was exported is corrupted. Something that supports that theory is that the v0.5.0-alpha.10 native client for Linux fails to load the model and exits with a RuntimeError. where the Trace-back points to `ds = Model(args.model, N_FEATURES, N_CONTEXT, args.alphabet, BEAM_WIDTH)`. Is there any other pre-trained model available that I can use for the newest version? Training is nearly impossible on my Laptop, so I rely on pre-trained models to properly test the bindings. It does not need to be any good at all. It should just produce some output so I can test the bindings.

For that kind of usage, we would recommend training a LDC93S1 model, it's a one-sample test set, it will make you able to work efficiently.

**Comment by GommeAntiLegit** [comment]
I'm currently working on the CUDA support for the bindings. Transcription of the audio is successful with both the CPU and CUDA versions of the DeepSpeech shared library. As speed of transcription is about the same, is there any way to verify that my GPU is used?

**Comment by lissyx** [comment]
> As speed of transcription is about the same, is there any way to verify that my GPU is used?

Yeah, TensorFlow should output some infos about GPUs. You should also get more informations using `TF_CPP_MIN_LOG_LEVEL` env variable, look into the output of it for `device:GPU`.

**Comment by GommeAntiLegit** [comment]
Cuda support is now present and works. Due to a bug, I was loading the CPU library, so that is why my GPU was not used. I wonder why... Duh... Many errors that could come up on Cuda runtime initialization are now wrapped with exceptions to make sure errors are human-readable (No random error codes etc).
Cuda Exceptions:
```
CudaInitializationError
CudaInsufficientDriverError
CudaInvalidDeviceError
CudaInvalidValueError
CudaNoDeviceError
CudaNotPermittedError
NoCudaJNIException
UnknownComputingModeError
UnknownCudaError
```
To make that possible I had to provide access to a few native Cuda functions (not accessible from outside the `Cuda` class):
`cudaGetDevice` [See documentation](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g80861db2ce7c29b6e8055af8ae01bc78)
`cudaGetDeviceProperties`  [See documentation](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g1bf9d625a931d657e08db2b4391170f0)
That means that there are now two build configurations for the `libdeepspeech-jni.so` library: 
`CUDA` and `CPU`.
Because of that, there is now one public function that throws a `NoCudaJNIException` when invoked with a CPU `libdeepspeech-jni.so` library loaded. I don't think it is necessary to have two different versions of the bindings. Just don't use `Cuda.getDevice()` when using the CPU version.

Cuda enabled `libdeepspeech.so`s should only be loaded in combination with Cuda enabled `libdeepspeech.so`s.
If the JNI library does not match the configuration specified in the java application, a `WrongJNILibraryException` is thrown.
This error is only possible because there is a function in `libdeepspeech-jni.so` that returns an integer representing the configuration.
But that is not possible for the `libdeepspeech.so` library. You just have to assume it's the correct configuration. Every other way of checking this would only be a cheap hack (eg. checking if certain functions are defined)
May I request this kind of function to be added?

eg: `int DS_GetConfiguration()`
returns 0 for `CPU,` 1 for `CUDA,` etc.

**Comment by lissyx** [comment]
@GommeAntiLegit Honestly, I fear that's a lot of over-engineering work.

**Comment by lissyx** [comment]
@GommeAntiLegit Everywhere else, we assume people won't mix incompatible things together, I feel it's more important right now to improve the quality of the Java bindings of all points you listed above instead of trying to craft too complicated solutions to a problem that barely exists: `libdeepspeech-jni.so` should load `libdeepspeech.so`, that's it. If people improperly package stuff, that's another issue, but we likely don't want to clutter the code with that, at least for now.

**Comment by GommeAntiLegit** [comment]
The bindings are now finished and the pull request has been made.

**Comment by lissyx** [comment]
> The bindings are now finished and the pull request has been made.

Issue should be closed once the matching PR lands :)
